alton/pause_event.py

```diff
@@ -0,0 +1,366 @@
+
+        Returns:
+            dict(pipeline_system: list()):
+                Dictionary with:
+                    keys: pipeline_system names
+                    values: lists of pipeline events as dicts
+        """
+        return {}
+
+
+class S3PauseEventOps(PauseEventOps):
+    """
+    Encapsulates S3 operations needed to pause/unpause pipeline systems and list pipeline statuses.
+    """
+    # Top-level directory name for all pause event files in bucket.
+    PAUSE_DIRECTORY = 'paused/'
+
+    # Subdirectory name holding the current pause events.
+    CURRENT_DIRECTORY = PAUSE_DIRECTORY + 'current/'
+
+    # Subdirectory holding all the historical pause events.
+    HISTORY_DIRECTORY = PAUSE_DIRECTORY + 'history/'
+
+    # Common time format to output/parse with strptime/strftime.
+    TIME_FORMAT = '%Y-%m-%d_%H:%M:%S'
+
+    def __init__(self, bucket_name, gocd_username, gocd_password, gocd_url):
+        self.s3_conn = boto.connect_s3()
+        # Get or create the specified bucket.
+        try:
+            self.pipeline_bucket = self.s3_conn.get_bucket(bucket_name)
+        except S3ResponseError as exc:
+            if exc.error_code == 'NoSuchBucket':
+                self.pipeline_bucket = self.s3_conn.create_bucket(bucket_name)
+        # Create a GoCD client for pausing/unpausing pipelines.
+        self.gocd_client = GoCDAPI(gocd_username, gocd_password, gocd_url)
+
+    def _create_s3_file(self, filepath, str_contents):
+        """
+        Create an S3 file at the filepath, writing the str_contents string to it.
+        """
+        s3_file = Key(self.pipeline_bucket, filepath)
+        s3_file.set_contents_from_string(str_contents)
+
+    def _get_current_pause_events(self, pipeline_system):
+        """
+        Returns the current pause status of one or all pipeline systems.
+
+        Arguments:
+            pipeline_system (str):
+                Pipeline system name for which to return status, e.g. edxapp, ecommerce, etc., None for all systems
+
+        Returns:
+            dict(pipeline_system: list()):
+                Dictionary with:
+                    keys: pipeline_system names
+                    values: lists of pipeline events as dicts
+        """
+        pause_status = defaultdict(list)
+        for key in bucket_lister(self.pipeline_bucket, prefix=self.CURRENT_DIRECTORY):
+            # Only read YAML files - at least, files with a .yml suffix.
+            if not key.name.endswith('.yml'):
+                continue
+            pause_data = yaml.safe_load(key.get_contents_as_string())
+            # If pipeline system was specified and this pause file isn't for that system, ignore it.
+            if pipeline_system and pipeline_system != pause_data['pipeline_system']:
+                continue
+            # Add the pause data from this file to the returned data.
+            pause_status[pause_data['pipeline_system']].append(pause_data)
+        return dict(pause_status)
```

Comment:
I was just cementing it into a regular dict at this point, since the code is finished adding to it. I'll try leaving it as a defaultdict.